src/storage/src/hummock/shared_buffer/shared_buffer_uploader.rs

```diff
@@ -140,66 +168,51 @@ impl SharedBufferUploader {
 
-        Ok(sync_size)
+        Ok(flush_size)
     }
 
     async fn handle(&mut self, item: SharedBufferUploaderItem) -> StorageResult<()> {
         match item {
-            SharedBufferUploaderItem::Batch(m) => {
+            SharedBufferUploaderItem::Batch(batch) => {
                 if let Some(detector) = &self.write_conflict_detector {
-                    detector.check_conflict_and_track_write_batch(&m.inner, m.epoch);
+                    detector.check_conflict_and_track_write_batch(&batch.inner, batch.epoch);
                 }
 
                 self.batches_to_upload
-                    .entry(m.epoch())
+                    .entry(batch.epoch())
                     .or_insert(Vec::new())
-                    .push(m);
-                Ok(())
+                    .push(batch);
             }
-            SharedBufferUploaderItem::Sync(sync_item) => {
-                let res = match sync_item.epoch {
-                    Some(e) => {
-                        // Sync a specific epoch
-                        self.sync(e).await
-                    }
-                    None => {
-                        // Sync all epochs
-                        let epochs = self.batches_to_upload.keys().copied().collect_vec();
-                        let mut res = Ok(0);
-                        let mut size_total: u64 = 0;
-
-                        for e in epochs {
-                            res = self.sync(e).await;
-                            if res.is_err() {
-                                break;
-                            }
-                            size_total += res.unwrap();
-                            res = Ok(size_total);
-                        }
-                        res
-                    }
-                };
-
-                if let Some(tx) = sync_item.notifier {
+            SharedBufferUploaderItem::Sync(epoch_notifier) => {
+                // Flush buffer to S3.
+                let res = self.flush(epoch_notifier.epoch, true).await;
+                // Notify sync notifier.
+                if let Some(tx) = epoch_notifier.notifier {
                     tx.send(res).map_err(|_| {
                         HummockError::shared_buffer_error(
                             "Failed to notify shared buffer sync because of send drop",
                         )
                     })?;
                 }
-                Ok(())
-            }
-            SharedBufferUploaderItem::Reset(epoch) => {
-                self.batches_to_upload.remove(&epoch);
-                Ok(())
             }
         }
+        Ok(())
     }
 
     pub async fn run(mut self) -> StorageResult<()> {
-        while let Some(m) = self.uploader_rx.recv().await {
-            if let Err(e) = self.handle(m).await {
-                return Err(e);
+        let mut ticker = tokio::time::interval(Duration::from_millis(
+            self.options.shared_buffer_upload_timeout as u64,
+        ));
```

Comment:
We may need to select a strategy for this `Interval`.

Reply:
You mean prevent from repeatedly selecting multiple ticks after a long flush?

Reply:
I do not understand why we need a interval flush after we have support flush by memory usage of shared-buffer

Reply:
@hzxa21

Reply:
> I do not understand why we need a interval flush after we have support flush by memory usage of shared-buffer

Time based batching has nothing to do with avoid OOM in shared buffer. Its main purpose is to reduce the latency of uploading a write batch.

Reply:
A second thought: what does the latency between a batch written into shared buffer and uploaded to S3 affect? 🤔

Reply:
the latency of a write batch = shared buffer wait time + S3 upload time. The latency can be reduced if shared buffer wait time is reduced.